Rom absolute tests

kevinhkhuynh opened the pull request:
1. Create rom absolute tests (run using -a option). This runs any tests starting with "absolute" in the script name.
2. Fix bug in Visit output path.

kevinhkhuynh commented:
I am having trouble getting some of the absolute offline tests to pass (l2 diff norm not below 1e-14) and absolute online tests to pass (speed up below 1). In any case, the infrastructure for absolute tests are there and anyone can make their own absolute tests by naming your script starting with "absolute" and following the example scripts I've put into the repo. Feel free to modify any of the thresholds of the tests I added.

kevinhkhuynh commented:
LGTM

dylan-copeland reviewed (approved):
Thanks Kevin. When someone has time, it will be good to run each absolute test manually to find perhaps better parameters to get reasonable accuracy and speed-up, but for now this is a good start.

chldkdtn commented:
> I am having trouble getting some of the absolute offline tests to pass (l2 diff norm not below 1e-14) and absolute online tests to pass (speed up below 1). In any case, the infrastructure for absolute tests are there and anyone can make their own absolute tests by naming your script starting with "absolute" and following the example scripts I've put into the repo. Feel free to modify any of the thresholds of the tests I added.

@kevinhkhuynh what are the tests you are struggling to get pass? Can you list them here?

kevinhkhuynh commented:
> 
> 
> > I am having trouble getting some of the absolute offline tests to pass (l2 diff norm not below 1e-14) and absolute online tests to pass (speed up below 1). In any case, the infrastructure for absolute tests are there and anyone can make their own absolute tests by naming your script starting with "absolute" and following the example scripts I've put into the repo. Feel free to modify any of the thresholds of the tests I added.
> 
> @kevinhkhuynh what are the tests you are struggling to get pass? Can you list them here?

I added four new tests and almost all of them fail due to not enough speed-up in online phase or l2 norm difference in the offline phase.

gresho
    $LAGHOS -p 4 -m data/square_gresho.mesh -rs 3 -ok 3 -ot 2 -tf 0.62 -s 7 -pa -offline -writesol -romsvds -sdim 10000
    $LAGHOS -p 4 -m data/square_gresho.mesh -rs 3 -ok 3 -ot 2 -tf 0.62 -s 7 -pa -online -rdimx 26 -rdimv 107 -rdime 90 -romhr -nsamx 3328 -nsamv 4802 -nsame 2304 -soldiff

sedov
    $LAGHOS -m data/cube01_hex.mesh -pt 211 -tf 0.4 -offline -writesol -romsvds
    $LAGHOS -m data/cube01_hex.mesh -pt 211 -tf 0.4 -online -rdimx 12 -rdimv 108 -rdime 27 -romhr -soldiff

taylor-green
 $LAGHOS -p 0 -rs 2 -iv -cfl 0.5 -tf 0.25 -pa -offline -romsvds -ef 0.9999 -writesol -romos -romsrhs -nwinsamp 10
 $LAGHOS -p 0 -rs 2 -iv -cfl 0.5 -tf 0.25 -pa -online -soldiff -romsvds -romos -romhr -romsrhs -romgs -sfacv 128 -sface 128 -twp twpTemp.csv -nwin 35

triple-point

0.1
0.2
0.3
0.4

 $LAGHOS -p 3 -m data/box01_hex.mesh -rs 1 -tf 0.04 -cfl 0.05 -pa -offline -writesol -romsvds -ef 0.9999 -nwin 4 -tw "$BASE_DIR"/tests/triple-point/triple-point-time-window.csv
 $LAGHOS -p 3 -m data/box01_hex.mesh -rs 1 -tf 0.04 -cfl 0.05 -pa -online -soldiff -nwin 4 -romhr -twp twpTemp.csv

kevinhkhuynh commented:
> 
> 
> > > I am having trouble getting some of the absolute offline tests to pass (l2 diff norm not below 1e-14) and absolute online tests to pass (speed up below 1). In any case, the infrastructure for absolute tests are there and anyone can make their own absolute tests by naming your script starting with "absolute" and following the example scripts I've put into the repo. Feel free to modify any of the thresholds of the tests I added.
> > 
> > 
> > @kevinhkhuynh what are the tests you are struggling to get pass? Can you list them here?
> 
> I added four new tests and almost all of them fail due to not enough speed-up in online phase or l2 norm difference in the offline phase.
> 
> gresho
> $LAGHOS -p 4 -m data/square_gresho.mesh -rs 3 -ok 3 -ot 2 -tf 0.62 -s 7 -pa -offline -writesol -romsvds -sdim 10000
> $LAGHOS -p 4 -m data/square_gresho.mesh -rs 3 -ok 3 -ot 2 -tf 0.62 -s 7 -pa -online -rdimx 26 -rdimv 107 -rdime 90 -romhr -nsamx 3328 -nsamv 4802 -nsame 2304 -soldiff
> 
> sedov
> $LAGHOS -m data/cube01_hex.mesh -pt 211 -tf 0.4 -offline -writesol -romsvds
> $LAGHOS -m data/cube01_hex.mesh -pt 211 -tf 0.4 -online -rdimx 12 -rdimv 108 -rdime 27 -romhr -soldiff
> 
> taylor-green
> $LAGHOS -p 0 -rs 2 -iv -cfl 0.5 -tf 0.25 -pa -offline -romsvds -ef 0.9999 -writesol -romos -romsrhs -nwinsamp 10
> $LAGHOS -p 0 -rs 2 -iv -cfl 0.5 -tf 0.25 -pa -online -soldiff -romsvds -romos -romhr -romsrhs -romgs -sfacv 128 -sface 128 -twp twpTemp.csv -nwin 35
> 
> triple-point
> 
> 0.1
> 0.2
> 0.3
> 0.4
> 
> $LAGHOS -p 3 -m data/box01_hex.mesh -rs 1 -tf 0.04 -cfl 0.05 -pa -offline -writesol -romsvds -ef 0.9999 -nwin 4 -tw "$BASE_DIR"/tests/triple-point/triple-point-time-window.csv
> $LAGHOS -p 3 -m data/box01_hex.mesh -rs 1 -tf 0.04 -cfl 0.05 -pa -online -soldiff -nwin 4 -romhr -twp twpTemp.csv

Perhaps the rostype load might fix some of them.

chldkdtn commented:
> > > > I am having trouble getting some of the absolute offline tests to pass (l2 diff norm not below 1e-14) and absolute online tests to pass (speed up below 1). In any case, the infrastructure for absolute tests are there and anyone can make their own absolute tests by naming your script starting with "absolute" and following the example scripts I've put into the repo. Feel free to modify any of the thresholds of the tests I added.
> > > 
> > > 
> > > @kevinhkhuynh what are the tests you are struggling to get pass? Can you list them here?
> > 
> > 
> > I added four new tests and almost all of them fail due to not enough speed-up in online phase or l2 norm difference in the offline phase.
> > gresho
> > $LAGHOS -p 4 -m data/square_gresho.mesh -rs 3 -ok 3 -ot 2 -tf 0.62 -s 7 -pa -offline -writesol -romsvds -sdim 10000
> > $LAGHOS -p 4 -m data/square_gresho.mesh -rs 3 -ok 3 -ot 2 -tf 0.62 -s 7 -pa -online -rdimx 26 -rdimv 107 -rdime 90 -romhr -nsamx 3328 -nsamv 4802 -nsame 2304 -soldiff
> > sedov
> > $LAGHOS -m data/cube01_hex.mesh -pt 211 -tf 0.4 -offline -writesol -romsvds
> > $LAGHOS -m data/cube01_hex.mesh -pt 211 -tf 0.4 -online -rdimx 12 -rdimv 108 -rdime 27 -romhr -soldiff
> > taylor-green
> > $LAGHOS -p 0 -rs 2 -iv -cfl 0.5 -tf 0.25 -pa -offline -romsvds -ef 0.9999 -writesol -romos -romsrhs -nwinsamp 10
> > $LAGHOS -p 0 -rs 2 -iv -cfl 0.5 -tf 0.25 -pa -online -soldiff -romsvds -romos -romhr -romsrhs -romgs -sfacv 128 -sface 128 -twp twpTemp.csv -nwin 35
> > triple-point
> > 0.1
> > 0.2
> > 0.3
> > 0.4
> > $LAGHOS -p 3 -m data/box01_hex.mesh -rs 1 -tf 0.04 -cfl 0.05 -pa -offline -writesol -romsvds -ef 0.9999 -nwin 4 -tw "$BASE_DIR"/tests/triple-point/triple-point-time-window.csv
> > $LAGHOS -p 3 -m data/box01_hex.mesh -rs 1 -tf 0.04 -cfl 0.05 -pa -online -soldiff -nwin 4 -romhr -twp twpTemp.csv
> 
> Perhaps the rostype load might fix some of them.

Would you try rostype load and see if that fixes the problems?

kevinhkhuynh commented:
> 
> 
> > > > > I am having trouble getting some of the absolute offline tests to pass (l2 diff norm not below 1e-14) and absolute online tests to pass (speed up below 1). In any case, the infrastructure for absolute tests are there and anyone can make their own absolute tests by naming your script starting with "absolute" and following the example scripts I've put into the repo. Feel free to modify any of the thresholds of the tests I added.
> > > > 
> > > > 
> > > > @kevinhkhuynh what are the tests you are struggling to get pass? Can you list them here?
> > > 
> > > 
> > > I added four new tests and almost all of them fail due to not enough speed-up in online phase or l2 norm difference in the offline phase.
> > > gresho
> > > $LAGHOS -p 4 -m data/square_gresho.mesh -rs 3 -ok 3 -ot 2 -tf 0.62 -s 7 -pa -offline -writesol -romsvds -sdim 10000
> > > $LAGHOS -p 4 -m data/square_gresho.mesh -rs 3 -ok 3 -ot 2 -tf 0.62 -s 7 -pa -online -rdimx 26 -rdimv 107 -rdime 90 -romhr -nsamx 3328 -nsamv 4802 -nsame 2304 -soldiff
> > > sedov
> > > $LAGHOS -m data/cube01_hex.mesh -pt 211 -tf 0.4 -offline -writesol -romsvds
> > > $LAGHOS -m data/cube01_hex.mesh -pt 211 -tf 0.4 -online -rdimx 12 -rdimv 108 -rdime 27 -romhr -soldiff
> > > taylor-green
> > > $LAGHOS -p 0 -rs 2 -iv -cfl 0.5 -tf 0.25 -pa -offline -romsvds -ef 0.9999 -writesol -romos -romsrhs -nwinsamp 10
> > > $LAGHOS -p 0 -rs 2 -iv -cfl 0.5 -tf 0.25 -pa -online -soldiff -romsvds -romos -romhr -romsrhs -romgs -sfacv 128 -sface 128 -twp twpTemp.csv -nwin 35
> > > triple-point
> > > 0.1
> > > 0.2
> > > 0.3
> > > 0.4
> > > $LAGHOS -p 3 -m data/box01_hex.mesh -rs 1 -tf 0.04 -cfl 0.05 -pa -offline -writesol -romsvds -ef 0.9999 -nwin 4 -tw "$BASE_DIR"/tests/triple-point/triple-point-time-window.csv
> > > $LAGHOS -p 3 -m data/box01_hex.mesh -rs 1 -tf 0.04 -cfl 0.05 -pa -online -soldiff -nwin 4 -romhr -twp twpTemp.csv
> > 
> > 
> > Perhaps the rostype load might fix some of them.
> 
> Would you try rostype load and see if that fixes the problems?

Oops, I was missing a "-p X" option on the baseline branch. The offline error is normally about 1e-10 which fails if we choose 1e-14, causing the offline test to fail. If we want to relax it to 1e-09, it should pass. And the online fails due to lack of speedup.

chldkdtn commented on rom/tests/gresho-vortices/absolute-gresho-vortices.sh:
```diff
@@ -1,16 +1,16 @@
     ;;
   2)
-    $LAGHOS -p 4 -m data/square_gresho.mesh -rs 3 -ok 3 -ot 2 -tf 0.62 -s 7 -pa -online -rdimx 26 -rdimv 107 -rdime 90 -romhr -nsamx 3328 -nsamv 4802 -nsame 2304 -soldiff
+    $LAGHOS -p 4 -m data/square_gresho.mesh -rs 3 -ok 3 -ot 2 -tf 0.62 -s 7 -pa -online -rdimx 26 -rdimv 107 -rdime 90 -romhr -nsamx 3328 -nsamv 4802 -nsame 2304 -soldiff -rostype load
```
It is hard to obtain a speed-up without time-windowing approach. Please see the following confluence page to set a problem up to get a speed-up:

https://myconfluence.llnl.gov/pages/viewpage.action?pageId=274661939

chldkdtn commented on rom/tests/sedov-blast/absolute-sedov-blast.sh:
```diff
@@ -8,10 +8,10 @@ absoluteRelErrorTolParallel="1e-1"
   2)
 
-    $LAGHOS -m data/cube01_hex.mesh -pt 211 -tf 0.4 -online -rdimx 12 -rdimv 108 -rdime 27 -romhr -soldiff
+    $LAGHOS -m data/cube01_hex.mesh -pt 211 -tf 0.4 -online -rdimx 12 -rdimv 108 -rdime 27 -romhr -soldiff -rostype load
```
Same here. A speed-up can be achieved by time-window approach. Please see the confluence page for sedov-blast problem.

chldkdtn reviewed (commented):
I am okay with 

> > > > > > I am having trouble getting some of the absolute offline tests to pass (l2 diff norm not below 1e-14) and absolute online tests to pass (speed up below 1). In any case, the infrastructure for absolute tests are there and anyone can make their own absolute tests by naming your script starting with "absolute" and following the example scripts I've put into the repo. Feel free to modify any of the thresholds of the tests I added.
> > > > > 
> > > > > 
> > > > > @kevinhkhuynh what are the tests you are struggling to get pass? Can you list them here?
> > > > 
> > > > 
> > > > I added four new tests and almost all of them fail due to not enough speed-up in online phase or l2 norm difference in the offline phase.
> > > > gresho
> > > > $LAGHOS -p 4 -m data/square_gresho.mesh -rs 3 -ok 3 -ot 2 -tf 0.62 -s 7 -pa -offline -writesol -romsvds -sdim 10000
> > > > $LAGHOS -p 4 -m data/square_gresho.mesh -rs 3 -ok 3 -ot 2 -tf 0.62 -s 7 -pa -online -rdimx 26 -rdimv 107 -rdime 90 -romhr -nsamx 3328 -nsamv 4802 -nsame 2304 -soldiff
> > > > sedov
> > > > $LAGHOS -m data/cube01_hex.mesh -pt 211 -tf 0.4 -offline -writesol -romsvds
> > > > $LAGHOS -m data/cube01_hex.mesh -pt 211 -tf 0.4 -online -rdimx 12 -rdimv 108 -rdime 27 -romhr -soldiff
> > > > taylor-green
> > > > $LAGHOS -p 0 -rs 2 -iv -cfl 0.5 -tf 0.25 -pa -offline -romsvds -ef 0.9999 -writesol -romos -romsrhs -nwinsamp 10
> > > > $LAGHOS -p 0 -rs 2 -iv -cfl 0.5 -tf 0.25 -pa -online -soldiff -romsvds -romos -romhr -romsrhs -romgs -sfacv 128 -sface 128 -twp twpTemp.csv -nwin 35
> > > > triple-point
> > > > 0.1
> > > > 0.2
> > > > 0.3
> > > > 0.4
> > > > $LAGHOS -p 3 -m data/box01_hex.mesh -rs 1 -tf 0.04 -cfl 0.05 -pa -offline -writesol -romsvds -ef 0.9999 -nwin 4 -tw "$BASE_DIR"/tests/triple-point/triple-point-time-window.csv
> > > > $LAGHOS -p 3 -m data/box01_hex.mesh -rs 1 -tf 0.04 -cfl 0.05 -pa -online -soldiff -nwin 4 -romhr -twp twpTemp.csv
> > > 
> > > 
> > > Perhaps the rostype load might fix some of them.
> > 
> > 
> > Would you try rostype load and see if that fixes the problems?
> 
> Oops, I was missing a "-p X" option on the baseline branch. The offline error is normally about 1e-10 which fails if we choose 1e-14, causing the offline test to fail. If we want to relax it to 1e-09, it should pass. And the online fails due to lack of speedup.

What are you comparing for offline error. the FOM in master with FOM in rom folder?

chldkdtn commented:
@kevinhkhuynh Please consult the confluence page to get a problem setup that gives you a speedup for each problem. Note that time windowing approach is needed for a speed-up.

kevinhkhuynh commented:
> 
> 
> I am okay with
> 
> > > > > > > I am having trouble getting some of the absolute offline tests to pass (l2 diff norm not below 1e-14) and absolute online tests to pass (speed up below 1). In any case, the infrastructure for absolute tests are there and anyone can make their own absolute tests by naming your script starting with "absolute" and following the example scripts I've put into the repo. Feel free to modify any of the thresholds of the tests I added.
> > > > > > 
> > > > > > 
> > > > > > @kevinhkhuynh what are the tests you are struggling to get pass? Can you list them here?
> > > > > 
> > > > > 
> > > > > I added four new tests and almost all of them fail due to not enough speed-up in online phase or l2 norm difference in the offline phase.
> > > > > gresho
> > > > > $LAGHOS -p 4 -m data/square_gresho.mesh -rs 3 -ok 3 -ot 2 -tf 0.62 -s 7 -pa -offline -writesol -romsvds -sdim 10000
> > > > > $LAGHOS -p 4 -m data/square_gresho.mesh -rs 3 -ok 3 -ot 2 -tf 0.62 -s 7 -pa -online -rdimx 26 -rdimv 107 -rdime 90 -romhr -nsamx 3328 -nsamv 4802 -nsame 2304 -soldiff
> > > > > sedov
> > > > > $LAGHOS -m data/cube01_hex.mesh -pt 211 -tf 0.4 -offline -writesol -romsvds
> > > > > $LAGHOS -m data/cube01_hex.mesh -pt 211 -tf 0.4 -online -rdimx 12 -rdimv 108 -rdime 27 -romhr -soldiff
> > > > > taylor-green
> > > > > $LAGHOS -p 0 -rs 2 -iv -cfl 0.5 -tf 0.25 -pa -offline -romsvds -ef 0.9999 -writesol -romos -romsrhs -nwinsamp 10
> > > > > $LAGHOS -p 0 -rs 2 -iv -cfl 0.5 -tf 0.25 -pa -online -soldiff -romsvds -romos -romhr -romsrhs -romgs -sfacv 128 -sface 128 -twp twpTemp.csv -nwin 35
> > > > > triple-point
> > > > > 0.1
> > > > > 0.2
> > > > > 0.3
> > > > > 0.4
> > > > > $LAGHOS -p 3 -m data/box01_hex.mesh -rs 1 -tf 0.04 -cfl 0.05 -pa -offline -writesol -romsvds -ef 0.9999 -nwin 4 -tw "$BASE_DIR"/tests/triple-point/triple-point-time-window.csv
> > > > > $LAGHOS -p 3 -m data/box01_hex.mesh -rs 1 -tf 0.04 -cfl 0.05 -pa -online -soldiff -nwin 4 -romhr -twp twpTemp.csv
> > > > 
> > > > 
> > > > Perhaps the rostype load might fix some of them.
> > > 
> > > 
> > > Would you try rostype load and see if that fixes the problems?
> > 
> > 
> > Oops, I was missing a "-p X" option on the baseline branch. The offline error is normally about 1e-10 which fails if we choose 1e-14, causing the offline test to fail. If we want to relax it to 1e-09, it should pass. And the online fails due to lack of speedup.
> 
> What are you comparing for offline error. the FOM in master with FOM in rom folder?

Yes FOM in master branch is compared with FOM in rom folder

chldkdtn commented:
@kevinhkhuynh is every absolute test passing now?

kevinhkhuynh commented:
> 
> 
> @kevinhkhuynh is every absolute test passing now?

Every test is passing except for absolute-taylor-green-offline and absolute-taylor-green-offline-parallel. I tried many different version of the taylor-green experiments on the Confluence, but unfortunately, it seems like there are some differences in the FOM solution between rom branch and the master branch, leading to an l2 diff norm that is greater than 1e-7 for taylor-green. The current test has an l2 diff norm of 20.0646 between the FOM solution of the rom branch and the master branch.

kevinhkhuynh commented:
@chldkdtn @dylan-copeland @siuwuncheung The differences between the offline phases from master and ROM are due to the many commits that were made in March on the master branch. The rom branch is fine and has the same offline results from when it first diverged from the master branch. So this is not due to a bug but due to rom not being in sync with master. How do we want to handle this? We can do a slow step-by-step rebase, but it might be difficult. I'm not sure of a better way since so many changes were made in rom since then.

chldkdtn commented on rom/tests/basisComparator.cpp:
```diff
@@ -62,21 +67,28 @@ between the basis matrices in the following files: " << baselineFile << " and "
         // Test whether l2 norm is smaller than error bound
-        if (matrixNormL2 > errorBound) {
-            cerr << "matrixNormL2 = " << matrixNormL2 << endl;
+        if (error > errorBound) {
```
Although it is not likely there will be case when baselineNormL2 is zero. Then error is not defined. If indeed the baselineNormL2 is zero, then we should look into the absolute difference as you did before. Would you add the case that baselineNormL2 is zero?

chldkdtn commented on rom/tests/solutionComparator.cpp:
```diff
@@ -70,11 +70,14 @@ void compareSolutions(string &baselineFile, string &targetFile, double errorBoun
     }
+    double baselineNormL2 = baseline.Norml2();
     double diffNormL2 = diff.Norml2();
+    double error = diffNormL2 / baselineNormL2;
```
same here

chldkdtn commented on rom/tests/fileComparator.cpp:
```diff
@@ -27,12 +27,14 @@ void compareFiles(ifstream &baselineFile, ifstream &targetFile, double errorBoun
+        stripped = targetLine.substr(posOfData != string::npos ? posOfData : 0);
         targetNum = stod(stripped);
+        double diff = baselineNum - targetNum;
+        double error = abs(baselineNum - targetNum) / baselineNum;
```
same here.
